R/reactives.R

```diff
@@ -38,6 +38,224 @@ Dependents <- R6Class(
+
+      if (private$frozen)
+        reactiveStop()
+
+      private$value
+    },
+    set = function(value) {
+      if (identical(private$value, value)) {
+        return(invisible(FALSE))
+      }
+      private$value <- value
+      .graphValueChange(private$label, value)
+      private$dependents$invalidate()
+      invisible(TRUE)
+    },
+    freeze = function(session = getDefaultReactiveDomain()) {
+      if (is.null(session)) {
+        stop("Can't freeze a reactiveVal without a reactive domain")
+      }
+      session$onFlushed(function() {
+        self$thaw()
+      })
+      private$frozen <- TRUE
+    },
+    thaw = function() {
+      private$frozen <- FALSE
+    },
+    isFrozen = function() {
+      private$frozen
+    },
+    format = function(...) {
+      label <- capture.output(print(base::format(private$value, ...)))
+      if (length(label) == 1) {
+        paste0("reactiveVal: ", label)
+      } else {
+        c("reactiveVal:", label)
+      }
+    }
+  )
+)
+
+#' Create a (single) reactive value
+#'
+#' A \code{reactiveVal} is an object used for reading and writing a value, like
+#' a variable, but with special capabilities for reactive programming. When you
+#' read the value out of a \code{reactiveVal}, the calling reactive expression
+#' takes a dependency, and when you change the value, it notifies any reactives
+#' that previously depended on that value.
+#'
+#' \code{reactiveVal} is very similar to \code{\link{reactiveValues}}, except
+#' that the former is for a single reactive value (like a variable), whereas the
+#' latter lets you conveniently use multiple reactive values by name (like a
+#' named list of variables). For a one-off reactive value, it's more natural to
+#' use \code{reactiveVal}.
+#'
+#' @param value An optional initial value.
+#' @param label An optional label, for debugging purposes (see
+#'   \code{\link{showReactLog}}). If missing, a label will be automatically
+#'   created.
+#'
+#' @return A function. Call the function with no arguments to (reactively) read
+#'   the value; call the function with a single argument to set the value.
+#'
+#' @examples
+#'
+#' \dontrun{
+#'
+#' # Create the object by calling reactiveVal
+#' r <- reactiveVal()
+#'
+#' # Set the value by calling with an argument
+#' r(10)
+#'
+#' # Read the value by calling without arguments
+#' r()
+#'
+#' }
+#'
+#' ## Only run examples in interactive R sessions
+#' if (interactive()) {
+#'
+#' ui <- fluidPage(
+#'   actionButton("minus", "-1"),
+#'   actionButton("plus", "+1"),
+#'   br(),
+#'   textOutput("value")
+#' )
+#'
+#' server <- function(input, output, session) {
+#'   value <- reactiveVal(0)
+#'
+#'   observeEvent(input$minus, {
+#'     newValue <- value() - 1
+#'     value(newValue)
+#'   })
+#'
+#'   observeEvent(input$plus, {
+#'     newValue <- value() + 1
+#'     value(newValue)
+#'   })
+#'
+#'   output$value <- renderText({
+#'     value()
+#'   })
+#' }
+#'
```

Comment:
It may be helpful to use this opportunity to compare and contrast `reactiveVal()` and `reactiveValues()`?

E.g. you could change the server function to this:
```r
# The comments below accomplish the same behavior using `reactiveValues()`
server <- function(input, output, session) {
    value <- reactiveVal(0)     # rv <- reactiveValues(value = 0)
    
    observeEvent(input$minus, {
        newValue <- value() - 1 #
        value(newValue)         # rv$value <- rv$value - 1
    })
    
    observeEvent(input$plus, {
        newValue <- value() + 1 #
        value(newValue)         # rv$value <- rv$value + 1
    })
    
    output$value <- renderText({
        value()                 # rv$value
    })
}
```